websocket_client.py

```diff
@@ -0,0 +1,463 @@
+        )
+        self._sink = LocalAudioSink()
+
+    async def start(self):
+        logging.info(f"Connecting to {self._url}")
+        self._socket = await ws_client.connect(self._url)
+        self._receive_task = asyncio.create_task(self.socket_receive(self._socket))
+
+    async def socket_receive(self, socket: ws_client.ClientConnection):
+        try:
+            async for message in socket:
+                await self._on_socket_message(message)
+        except asyncio.CancelledError:
+            logging.info("socket cancelled")
+        except ws_exceptions.ConnectionClosedOK:
+            logging.info("socket closed ok")
+        except ws_exceptions.ConnectionClosedError as e:
+            self.emit("error", e)
+            return
+        logging.info("socket receive done")
+        self.emit("ended")
+
+    async def stop(self):
+        logging.info("Stopping...")
+        await _async_close(
+            self._sink.close(),
+            self._socket.close() if self._socket else None,
+            _async_cancel(self._send_audio_task, self._receive_task),
+        )
+        if self._state != "idle":
+            self._state = "idle"
+            self.emit("state", "idle")
+
+    async def _on_socket_message(self, payload: str | bytes):
+        if isinstance(payload, bytes):
+            self._sink.write(payload)
+            return
+        elif isinstance(payload, str):
+            msg = json.loads(payload)
+            await self.handle_data_message(msg)
+
+    async def handle_data_message(self, msg: dict[str, Any]):
+        match msg["type"]:
+            case "playback_clear_buffer":
+                self._sink.drop_buffer()
+            case "state":
+                if msg["state"] != self._state:
+                    self._state = msg["state"]
+                    self.emit("state", msg["state"])
+            case "transcript":
+                # This is lazy handling of transcripts. See the WebRTC client SDKs
+                # for a more robust implementation.
+                if msg["role"] != "agent":
+                    return  # Ignore user transcripts
+                if msg.get("text", None):
+                    self._pending_output = msg["text"]
+                    self.emit("output", msg["text"], msg["final"])
+                else:
+                    self._pending_output += msg.get("delta", "")
+                    self.emit("output", self._pending_output, msg["final"])
+                if msg["final"]:
+                    self._pending_output = ""
+            case "client_tool_invocation":
+                await self._handle_client_tool_call(
+                    msg["toolName"], msg["invocationId"], msg["parameters"]
+                )
+            case "debug":
+                logging.info(f"debug: {msg['message']}")
+            case _:
+                logging.warning(f"Unhandled message type: {msg['type']}")
+
+    async def _handle_client_tool_call(
```

Comment:
"""
        Process client tool invocations.
  
        See https://docs.ultravox.ai/essentials/tools for more on tools.
        
        """